android/envoy/src/main/kotlin/org/greatfire/envoy/EnvoyConnectWorker.kt

```diff
@@ -1,62 +1,118 @@
 package org.greatfire.envoy
 
 import android.content.Context
+import android.content.SharedPreferences
 import android.util.Log
+import androidx.preference.PreferenceManager
 import androidx.work.CoroutineWorker
 import androidx.work.WorkerParameters
 import java.net.URI
 import kotlinx.coroutines.*
+import org.greatfire.envoy.NetworkIntentService.Companion
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicLong
 
 /*
     Establish a connection to an Envoy Proxy
 */
 class EnvoyConnectWorker(
-    val context: Context, val params: WorkerParameters
+    val context: Context, val params: WorkerParameters, val callback: EnvoyTestCallback
 ) : CoroutineWorker(context, params) {
 
+    // MNB read about workers
+
     val tests = EnvoyConnectionTests()
 
     companion object {
         private const val TAG = "EnvoyConnectWorker"
+
+        private const val TIME_LIMIT = 60000 // make configurable?
+        private const val ONE_HOUR_MS = 3600000
+        private const val ONE_DAY_MS = 86400000
+        private const val ONE_WEEK_MS = 604800000
+        private const val TIME_SUFFIX = "_time"
+        private const val COUNT_SUFFIX = "_count"
     }
 
     // this ArrayDeque is our working copy of the tests
     private val envoyTests = ArrayDeque<EnvoyTest>()
     private val jobs = mutableListOf<Job>()
 
+    // simple counters to infer status
+    private var testCount = AtomicInteger()
+    private var blockedCount = AtomicInteger()
+    private var failedCount = AtomicInteger()
+
+    private var startTime = AtomicLong()
+
+    private var foundUrl = AtomicBoolean()
```

Comment:
This seems to track the value I was tracking in `EnvoyNetworking.activeUrl`, well I guess it's the value of `EnovyNetorking.activeURL != ""`